File: DSandAlgos/reverseLinkedList2.js
// Given the head of a singly linked list and two integers left and right where left <= right, reverse the nodes of the list from position left to position right, and return the reversed list.

function ListNode(val, next) {
    this.val = val === undefined ? 0 : val
    this.next = next === undefined? null : next
}

function reverseInnerLinkedList(head, left, right) {
    let current = head, previous = null, start, end, i=0
    while (i<left-1) {
        previous = current
        current = current.next
        i++
    }

    let j = 0
    start = previous
    previous = null
    end = current
    while (current && j<right-left+1) {
        let next = current.next
        current.next = previous
        previous = current
        current = next
        j++
    }

    if (start) {
        start.next = previous
    } else {
        head = previous
    }

    end.next = current
    return head
}

File: DSandAlgos/ReverseLinkedList/reverseLinkedList2.py
from contextlib import nullcontext


class ListNode:
    def init(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def reverse_linked_list_2(self, head= Optional[ListNode], left= int, right= int)->Optional[ListNode]:
        i, current, previous, start, end = 0, head, None, None, None

        while i<left-1:
            previous = current
            current = current.next
            i+=1

        j = 0
        start = previous
        previous = None
        end = current

        while current and j<right-left+1:
            next = current.next
            current.next = previous
            previous = current
            current = next
            j+=1
        
        if start != None:
            start.next = previous
        else:
            head = previous

        end.next = current
        return head